glean-core/src/upload/mod.rs

```diff
@@ -30,48 +30,77 @@ use util::process_pings_dir;
 mod request;
 mod util;
 
+/// When asking for the next ping request to upload,
+/// the requester may receive one out of three possible tasks.
+///
+/// This enum is a representation of these possibilities.
+#[derive(PartialEq, Debug)]
+pub enum PingUploadTask {
+    /// A PingRequest popped from the front of the queue.
+    /// See [`PingRequest`](struct.PingRequest.html) for more information.
+    Upload(PingRequest),
+    /// A flag signaling that the pending pings directories are not done being processed,
+    /// thus the requester should wait and come back later.
+    Wait,
+    /// A flag signaling that the pending pings queue is empty and requester is done.
+    Done,
+}
+
 /// Manages the pending pings queue and directory.
 #[derive(Debug)]
 pub struct PingUploadManager {
     /// A FIFO queue storing a `PingRequest` for each pending ping.
     queue: Arc<RwLock<VecDeque<PingRequest>>>,
-    /// A flag, signaling if we are done processing the pending pings directories.
+    /// A flag signaling if we are done processing the pending pings directories.
+    ///
+    /// This flags value does not mean that the processing of the directory was successful,
+    /// only that it happened or not yet.
     processed_pending_pings: Arc<AtomicBool>,
 }
 
 impl PingUploadManager {
     /// Create a new PingUploadManager.
     ///
-    /// This will spawn a new thread and processes the pending pings folder
-    /// filling up the queue with whatever pings are in there
-    /// ordered by oldest to newest file modified.
+    /// Spawns a new thread and processes the pending pings directory,
```

Comment:
I am sorry for the back and forth, but I didn't think it made sense anymore to not do this on init once I changed the flag as per @badboy 's review. I moved it back to the init, which ultimately I think is better, because the caller doesn't need to init and also call the `process_pings_dir` function.

The only bad thing about this is that it required some _hacks_ to get the tests to work everytime, but I don't think they are too bad. Let me know your thoughts.